Add support for automation-hub authentication to ansible-galaxy

alikins opened the pull request:
And pass it to GalaxyAPI() from cli.

Add a KeycloakToken for use with keycloak sso setups
ie, sso.redhat.com

Fixes: ansible/galaxy-dev#96

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ansible-galaxy

alikins committed:
Add 'auth_url' to galaxy server config stanza

And pass it to GalaxyAPI() from cli.

Add a KeycloakToken for use with keycloak sso setups
ie, sso.redhat.com

Fixes: ansible/galaxy-dev#96

alikins commented:
pinging @AlanCoding @ironfroggy @chouseknecht

alikins committed:
Pop 'auth_url' from server_opts for GalaxyAPI

GalaxyAPI() doesn't need an 'auth_url' arg, so
pop it from server_options before server_options
is sent to GalaxyAPI as kwargs.

alikins committed:
Use basic auth if server def has username

Even if it also has an 'auth_url' ala automation-hub

alikins committed:
Specify auth is required for v3 requests.

alikins committed:
Move token serialization to Token classes

alikins committed:
Handle token auth or basic auth for autohub

alikins commented:
An example ansible-cfg for using offline tokens for automation_hub_ci:

``` ini
[galaxy]
server_list = automation_hub_ci

[galaxy_server.automation_hub_ci]
url=https://ci.cloud.redhat.com/api/automation-hub/
auth_url=https://sso.qa.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token
# Get the "offline token" from https://ci.cloud.redhat.com/beta/ansible/automation-hub/token and use for "token" value
token=<the offline token JWT blob from above url>

```

Example using user/pass basic http auth (not recomended):

``` ini
[galaxy]
server_list = automation_hub_ci

[galaxy_server.automation_hub_ci]
url=https://ci.cloud.redhat.com/api/automation-hub/
username=jethro_null
password=aqua1ung
```

alikins committed:
Add changelog framgement for automation-hub ansible-galaxy changes

alikins committed:
Add 'auth_url' to galaxy config documentation.

And an example of using it in a example config
for Automation Hub

samccann reviewed (approved):
Docs portion lgtm

AlanCoding reviewed (approved):
my testing seems to verify that this gets it working with username/password auth

AlanCoding commented:
Only nit I have with this is that now it generates confusing outputs for collections that cannot be found:

```
Process install dependency map
Initial connection to galaxy_server: https://ci.cloud.redhat.com/api/automation-hub/
Calling Galaxy at https://ci.cloud.redhat.com/api/automation-hub/api
Processing requirement collection 'chrismeyersfsufff.test_thingsfff'
Collection requirement 'chrismeyersfsufff.test_thingsfff' is the name of a collection
Calling Galaxy at https://ci.cloud.redhat.com/api/automation-hub/api
Found API version 'v3' with Galaxy server primary_galaxy (https://ci.cloud.redhat.com/api/automation-hub/)
Calling Galaxy at https://ci.cloud.redhat.com/api/automation-hub/v3/collections/chrismeyersfsufff/test_thingsfff/versions
Collection 'chrismeyersfsufff.test_thingsfff' obtained from server primary_galaxy https://ci.cloud.redhat.com/api/automation-hub/
ERROR! Cannot meet requirement * for dependency chrismeyersfsufff.test_thingsfff from source 'https://ci.cloud.redhat.com/api/automation-hub/'. Available versions before last requirement added: 
Requirements from:
	base - 'chrismeyersfsufff.test_thingsfff:*'
```

It says a thing clearly incorrect at "Collection 'chrismeyersfsufff.test_thingsfff' obtained from server primary_galaxy https://ci.cloud.redhat.com/api/automation-hub/". The name "primary_galaxy" is what I called the server.

I'm not totally sure what's going on there, other than that it's failing at the wrong stage of processing things. This should presumably go as a new issue here.

alikins committed:
Pass in validate_certs to KeycloakInit()

Instead of pulling in context

alikins committed:
For paginated requests, make iters use same auth_required

alikins committed:
Update galaxy api unit tests

Mostly changes related to v3 related auth changes.

chouseknecht commented on lib/ansible/galaxy/api.py:
```diff
@@ -500,13 +500,20 @@ def get_collection_version_metadata(self, namespace, name, version):
-        api_path = self.available_api_versions.get('v3', self.available_api_versions.get('v2'))
+        if 'v3' in self.available_api_versions:
+            api_path = self.available_api_versions['v3']
+            auth_required = True
```
Just because it's `v3` doesn't mean auth is required. For Automation Hub, yes, `v3` requires auth to access metadata. In the future when community Galaxy has a `v3`, I suspect auth won't be required.

alikins replied:
Agreed. But at the moment, there is no other mechanism to determine if auth is required.

chouseknecht commented on lib/ansible/galaxy/collection.py:
```diff
@@ -153,7 +153,14 @@ def install(self, path, b_temp_path):
+            # for v2 api, we don't require or add auth
+            # for v3 api, we require auth and add it, but not for download redirects
+            if 'v3' in self.api.available_api_versions:
+                auth_required = True
```
Same as above. Don't think we should assume that `v3` requires authorization.

alikins committed:
Mock _avail_api_versions to avoid network call

alikins committed:
Add BasicAuthToken

alikins committed:
Stop assuming v3 requires auth for GETs

Tweak _add_auth_api() to allow adding any sort of
auth including no auth unless 'required' is true.

Stop requiring auth for most things at the client
level. Let the server decide.

Auth is setup and used if it is found in the config.

alikins committed:
Let BasicAuthToken password arg be optional

alikins committed:
Fix unit test (no extra unauthed /api call)

alikins committed:
Rm unneeded style change / diff minimizing

chouseknecht commented:
Failures in shippable are related to Azure and MySQL and not anything to do with code changes made here. Merging.
